### PR TITLE
Obsolete: Support aggregation functions and arithmetic RTS-848

### PR DESCRIPTION
**Pull request by andytill**
Support aggregation functions and arithmetic. For example.

``` sql
COUNT(foo) + 1
```

And

``` sql
COUNT(foo) + COUNT(bar)
```

Clauses like.

``` sql
COUNT(COUNT(foo))
```

Will not be supported.  Depends on riak_ql PR https://github.com/basho/riak_ql/pull/90 && https://github.com/basho/riak_ql/pull/93

### CODE REVIEW

File: src/riak_kv_qry_compiler.erl [hunk 1]
```diff
@@ -207,52 +246,44 @@ compile_select_col(DDL, {{window_agg_fn, FnName}, [FnArg1]}) ->
...
             {ColTypes2, IsValid2, Compiled_arg1} =
                 compile_select_col_stateless(DDL, FnArg1),
-            %% all the windows agg fns so far are arity of 1
-            %% which we have forced in this clause by matching on a single argument in the
-            %% function head
             FnSig = riak_ql_window_agg_fns:get_arity_and_type_sig(FnName),
             {IsValid, ColRet} = check_types(FnName, FnSig, ColTypes2),
 
             IsValid3 = merge_validation(IsValid2, IsValid),
+            % all the windows agg fns so far are arity of 1
```
Comment (hmmr) [inline src/riak_kv_qry_compiler.erl:268]: On behalf of all cat^W emacs people, please keep this comment block marked with double `%`. Autoindenting this block in Emacs will move it far to the right. (Ditto for other blocks elsewhere.)

File: src/riak_kv_qry_compiler.erl [hunk 2]
```diff
@@ -319,13 +374,13 @@ merge_validation({error, E1}, {error, E2}) -> {error, E1 ++ E2}.
 
 %%
 compile_select_col_stateless2('+', A, B) ->
-    fun(Row) -> A(Row) + B(Row) end;
+    fun(Row, State) -> A(Row, State) + B(Row, State) end;
```
Comment (hmmr) [inline src/riak_kv_qry_compiler.erl]: A `State` parameter in a function having '_stateless' in its name?

File: src/riak_kv_qry_compiler.erl [hunk 3]
```diff
@@ -1640,6 +1672,39 @@ basic_select_arith_1_test() ->
        Sel
       ).
 
+varchar_literal_test() ->
+    {ok, Rec} = get_query("SELECT 'hello' from mytab"),
```
Comment (hmmr) [inline src/riak_kv_qry_compiler.erl:1705]: I note that under normal circumstances, validation of a SELECT query without WHERE clause will not succeed. Is there a plan to allow WHERE-less queries with a static expression not referencing any columns?

(I understand these tests only deal with parts of DDL structure, so the above question is asked in a broader context.)

File: src/riak_kv_qry_compiler.erl (hunk 1 above)
Comment (andytill) [inline src/riak_kv_qry_compiler.erl:268]: I wondered why my comments kept getting double indented in other people's commits.

File: src/riak_kv_qry_compiler.erl (hunk 3 above)
Comment (andytill) [inline src/riak_kv_qry_compiler.erl:1705]: This is just to minimise test data. I'm not sure about plans to do things like `SELECT 2 FROM dual`, maybe if there is ever a need for sub queries.

File: src/riak_kv_qry_compiler.erl (hunk 2 above)
Comment (andytill) [inline src/riak_kv_qry_compiler.erl]: That variable is more like a `ColValue` than a state, or do you think there is a better name? IMHO the function is correctly named as it purposely does not handle aggregate functions.

File: src/riak_kv_qry_compiler.erl [hunk 4]
```diff
@@ -22,11 +22,9 @@
...
 -module(riak_kv_qry_compiler).
 
--export([
-         compile/3,
-         run_select/2,
-         run_select/3  %% what is this for?
-        ]).
+-export([compile/3]).
+-export([finalise_aggregate/2]).
+-export([run_select/2, run_select/3]).
```
Comment (hazen) [inline src/riak_kv_qry_compiler.erl:28]: Are multiple exports just stylistic preference?

**Comment by hazen** [comment]
All unit tests pass with the QL `at/aggregations_fns_and_arithmetic2` branch

File: src/riak_kv_qry_compiler.erl [hunk 5]
```diff
@@ -113,6 +128,14 @@ prepend_select_columns([_|_] = MultiCols, Acc) ->
 prepend_select_columns(V, Acc) ->
     [V | Acc].
 
+%% copy pasta mapfoldl from lists.erl, otherwise this reports
+%% a warning in dialyzer!
+my_mapfoldl(F, Accu0, [Hd|Tail]) ->
```
Comment (hazen) [inline src/riak_kv_qry_compiler.erl:133]: Really?  You can't use `lists:mapfoldl`? This is the dialyzer bug mentioned the other day?

Comment (lehoff) [inline src/riak_kv_qry_compiler.erl:133]: It is the dialyzer potential bug - I have asked Andrei to create a Jira ticket for this or add to the one Andy created.

File: src/riak_kv_qry_compiler.erl (hunk 4 above)
Comment (andytill) [inline src/riak_kv_qry_compiler.erl:28]: It makes it very easy to sort alphabetically and group functions with the same name but different arity.

**Pull request closed without merging**